brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+                    scan_onsets=None, coords=None, inten=None,
+                    inten_weight=0.0):
+        """ The major utility of fitting Bayesian RSA.
+        """
+        GP_inten = self.GP_inten
+        GP_space = self.GP_space
+        rank = self.rank
+        n_V = np.size(Y, axis=1)
+        n_T = np.size(Y, axis=0)
+        n_C = np.size(design, axis=1)
+        l_idx = np.tril_indices(n_C)
+
+        t_start = time.time()
+
+        if rank is not None:
+            # The rank of covariance matrix is specified
+            idx_rank = np.where(l_idx[1] < rank)
+            l_idx = (l_idx[0][idx_rank], l_idx[1][idx_rank])
+            logger.debug('Using the rank specified by the user: '
+                         '{}'.format(rank))
+        else:
+            rank = n_C
+            # if not specified, we assume you want to
+            # estimate a full rank matrix
+            logger.debug('Please be aware that you did not specify the rank'
+                         ' of covariance matrix you want to estimate.'
+                         'I will assume that the covariance matrix shared '
+                         'among voxels is of full rank.'
+                         'Rank = {}'.format(rank))
+            logger.debug('Please be aware that estimating a matrix of '
+                         'high rank can be veryy slow.'
+                         'If you have a good reason to specify a lower rank '
+                         'than the number of experiment conditions, do so.')
+
+        n_l = np.size(l_idx[0])  # the number of parameters for L
+
+        XTY, XTDY, XTFY, YTY_diag, YTDY_diag, YTFY_diag, XTX, XTDX, XTFX = \
+            self._prepare_data(design, Y, n_T, n_V, scan_onsets)
+
+        dist2, inten_diff2, space_smooth_range, inten_smooth_range,\
+            n_smooth = self.calc_dist2_GP(
+                coords=coords, inten=inten,
+                GP_space=GP_space, GP_inten=GP_inten)
+        current_GP = np.zeros(n_smooth)
+
+        beta_hat = np.linalg.lstsq(design, Y)[0]
+        residual = Y - np.dot(design, beta_hat)
+        # There are several possible ways of initialization.
+        # (1) start from the point estimation of covariance
+
+        # cov_point_est = np.cov(beta_hat)
+        # current_vec_U_chlsk_l_AR1 = \
+        #     np.linalg.cholesky(cov_point_est + \
+        #     np.eye(n_C) * 1e-6)[l_idx]
+
+        # We add a tiny diagonal element to the point
+        # estimation of covariance, just in case
+        # the user provides data in which
+        # n_V is smaller than n_C
+
+        # (2) start from identity matrix
+
+        # current_vec_U_chlsk_l_AR1 = np.eye(n_C)[l_idx]
+
+        # (3) random initialization
+
+        current_vec_U_chlsk_l_AR1 = np.random.randn(n_l)
+
+        param0 = np.zeros(n_l + 2)
+        param0_fitU = np.zeros(n_l + n_V)
+        param0_fitV = np.zeros(n_V + n_smooth - 1)
+
+        param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
```

Comment:
Here and elsewhere, would be good to have an array of indices rather than hardcode here -- makes it much harder to remember the organization of the parameters in the param vector.